{"cells":[{"cell_type":"markdown","source":["# `pandas` - Read data files"],"metadata":{}},{"cell_type":"markdown","source":["This notebook demonstrates the `read_csv`, `read_excel`, and `read_json` functions from pandas."],"metadata":{}},{"cell_type":"markdown","source":["## Contents\n1. Setup\n1. `read_csv` & fixing errors\n1. `read_excel`\n1. `read_json`"],"metadata":{}},{"cell_type":"markdown","source":["## 1. Setup"],"metadata":{}},{"cell_type":"markdown","source":["Load the `pandas` Python library."],"metadata":{}},{"cell_type":"code","source":["%python\nimport pandas as pd"],"metadata":{},"outputs":[],"execution_count":6},{"cell_type":"code","source":["%sh ls -hot /dbfs/mnt/datalab-datasets/538_data/mad-men"],"metadata":{},"outputs":[],"execution_count":7},{"cell_type":"code","source":["%sh ls -hot /dbfs/mnt/datalab-datasets/538_data/mad-men/*"],"metadata":{},"outputs":[],"execution_count":8},{"cell_type":"markdown","source":["Notice that the output above changes if the asterisk is removed/included in the filepath of the `ls` command. \n\nThe asterisk is required in order to produce a vector of full filepaths."],"metadata":{}},{"cell_type":"markdown","source":["## 2. `read_csv` & fixing errors"],"metadata":{}},{"cell_type":"markdown","source":["Attempt to read the `show-data.csv` file with the pandas `read_csv` function. \n\nIt produces an error, which we will fix below."],"metadata":{}},{"cell_type":"code","source":["%python # error encountered\nmad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv')\nmad_men_show_data_pdf"],"metadata":{},"outputs":[],"execution_count":12},{"cell_type":"markdown","source":["Notice the error above is also repeat below when simply displaying the file with the `cat` command."],"metadata":{}},{"cell_type":"code","source":["%sh # error encountered\ncat /dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv"],"metadata":{},"outputs":[],"execution_count":14},{"cell_type":"markdown","source":["In this case, when encountering an unknown error, I search for posts about the error on Google. \n\nSearch for: \"read_csv pandas UnicodeDecodeError utf-8 codec can't decode byte  in position invalid continuation byte\". \n\nNotice I added \"read_csv\" and \"pandas\" to the search string. \nI removed `0xd5` and `3032` as these are specific to our file and won't help the search find other posts that solve the same problem, but with a different file. \n\nThe search found: https://stackoverflow.com/questions/18171739/unicodedecodeerror-when-reading-csv-file-in-pandas-with-python\n\nThis links suggests trying `encoding = \"ISO-8859-1\"` as a parameter to `read_csv` (it works, see below)"],"metadata":{}},{"cell_type":"code","source":["%python\nmad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', \n                                    encoding = 'ISO-8859-1')\nmad_men_show_data_pdf.info()"],"metadata":{},"outputs":[],"execution_count":16},{"cell_type":"markdown","source":["The command did work. \n\nNotice the column names have non-alphabetic characters (including spaces).\nThe next step is to read in the columns with better names."],"metadata":{}},{"cell_type":"markdown","source":["The names parameter specifies the column names to use (in the dataframe returned by `read_csv`)."],"metadata":{}},{"cell_type":"code","source":["%python\nmad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', \n                                    names=['performer', 'show', 'show_start', 'show_end',\n                                           'status', 'char_end', 'years_since', 'lead', \n                                           'support', 'shows', 'score', 'score_y',\n                                           'lead_notes', 'support_notes', 'show_notes'],\n                                    header=0, \n                                    encoding = 'ISO-8859-1')\nmad_men_show_data_pdf.info()"],"metadata":{},"outputs":[],"execution_count":19},{"cell_type":"markdown","source":["Notice the column names are better, but that we needed to specify, with the `header` parameter, that row `0` containes the column names."],"metadata":{}},{"cell_type":"markdown","source":["The column types are in the right hand column of the output from the `info` method.\nIf a column is read in as a `float64` or `int64` then we can assume it was read correctly. \n\nTo check for columns that were not read correctly we only need to select columns with `object` type. \nThe next command does this."],"metadata":{}},{"cell_type":"code","source":["mad_men_show_data_pdf \\\n  .select_dtypes(include=['object']) \\\n  .head()"],"metadata":{},"outputs":[],"execution_count":22},{"cell_type":"markdown","source":["Notice that `show_end` and `score_y` should be numbers.\n\nThe following command specifies both `show_start` and `show_end` as integers. We leave `score_y` until later."],"metadata":{}},{"cell_type":"code","source":["%python # error encountered\nimport numpy as np\nmad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', \n                                    names=['performer', 'show', 'show_start', 'show_end',\n                                             'status', 'char_end', 'years_since', 'lead', \n                                             'support', 'shows', 'score', 'score_y',\n                                             'lead_notes', 'support_notes', 'show_notes'],\n                                    dtype={'show_start': np.int64, 'show_end': np.int64},\n                                    header=0, \n                                    encoding = 'ISO-8859-1')\nmad_men_show_data_pdf.info()"],"metadata":{},"outputs":[],"execution_count":24},{"cell_type":"markdown","source":["An error is encountered, indicating that `PRESENT` is an invalid value for an integer. \n\nSo we specify `PRESENT` to indicate a missing value. This should work."],"metadata":{}},{"cell_type":"code","source":["%python\nimport numpy as np\nmad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', \n                                    names=['performer', 'show', 'show_start', 'show_end',\n                                             'status', 'char_end', 'years_since', 'lead', \n                                             'support', 'shows', 'score', 'score_y',\n                                             'lead_notes', 'support_notes', 'show_notes'],\n                                    dtype={'show_start': np.int64, 'show_end': np.int64, \n                                           'score_y': np.float64},\n                                    header=0, \n                                    na_values=['PRESENT'],\n                                    encoding = 'ISO-8859-1')\nmad_men_show_data_pdf.info()"],"metadata":{},"outputs":[],"execution_count":26},{"cell_type":"markdown","source":["But this does not work. So we check Google again for the error."],"metadata":{}},{"cell_type":"markdown","source":["Searching for \"ValueError: Integer column has NA values in column 3\" yields this link:\n- https://stackoverflow.com/questions/21287624/convert-pandas-column-containing-nans-to-dtype-int\n\nThe post indicates that integers cannot have missing values in pandas and that they should be retyped to `np.float32`. \nBoth columns (`show_start` and `show_end`) where coding this way for consistency. \nIt's awkward, but it works."],"metadata":{}},{"cell_type":"code","source":["%python\nimport numpy as np\nmad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', \n                                    names=['performer', 'show', 'show_start', 'show_end',\n                                           'status', 'char_end', 'years_since', 'lead', \n                                           'support', 'shows', 'score', 'score_y',\n                                           'lead_notes', 'support_notes', 'show_notes'],\n                                    dtype={'show_start': np.float32, 'show_end': np.float32},\n                                    header=0, \n                                    na_values=['PRESENT'],\n                                    encoding = 'ISO-8859-1')\nmad_men_show_data_pdf.info()"],"metadata":{},"outputs":[],"execution_count":29},{"cell_type":"markdown","source":["Recall that`score_y` should be a `float`. The command below specifies this in the `dtype` parameter."],"metadata":{}},{"cell_type":"code","source":["%python\nimport numpy as np\nmad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', \n                                    names=['performer', 'show', 'show_start', 'show_end',\n                                           'status', 'char_end', 'years_since', 'lead', \n                                           'support', 'shows', 'score', 'score_y',\n                                           'lead_notes', 'support_notes', 'show_notes'],\n                                    dtype={'show_start': np.float32, \n                                           'show_end': np.float32,\n                                           'score_y': np.float32},\n                                    header=0, \n                                    na_values=['PRESENT'],\n                                    encoding = 'ISO-8859-1')\nmad_men_show_data_pdf.info()"],"metadata":{},"outputs":[],"execution_count":31},{"cell_type":"markdown","source":["The error indicates that `#DIV/0!` is a value in the `score_y` field. So this value is included in the `na_values` parameter (below)."],"metadata":{}},{"cell_type":"code","source":["%python\nimport numpy as np\nmad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', \n                                    names=['performer', 'show', 'show_start', 'show_end',\n                                           'status', 'char_end', 'years_since', 'lead', \n                                           'support', 'shows', 'score', 'score_y',\n                                           'lead_notes', 'support_notes', 'show_notes'],\n                                    dtype={'show_start': np.float32, \n                                           'show_end': np.float32,\n                                           'score_y': np.float32},\n                                    header=0, \n                                    na_values=['PRESENT','#DIV/0!'],\n                                    encoding = 'ISO-8859-1')\nmad_men_show_data_pdf.info()"],"metadata":{},"outputs":[],"execution_count":33},{"cell_type":"markdown","source":["Great. All columns are read with acceptable datatypes. The following command displays the first few records for `object` columns."],"metadata":{}},{"cell_type":"code","source":["%python\nimport numpy as np\nmad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', \n                                    names=['performer', 'show', 'show_start', 'show_end',\n                                           'status', 'char_end', 'years_since', 'lead', \n                                           'support', 'shows', 'score', 'score_y',\n                                           'lead_notes', 'support_notes', 'show_notes'],\n                                    dtype={'show_start': np.float32, \n                                           'show_end': np.float32,\n                                           'score_y': np.float32},\n                                    header=0, \n                                    na_values=['PRESENT','#DIV/0!'],\n                                    encoding = 'ISO-8859-1')\nmad_men_show_data_pdf \\\n  .select_dtypes(include=['object']) \\\n  .head(5)"],"metadata":{},"outputs":[],"execution_count":35},{"cell_type":"markdown","source":["One final modification is to check on the `status` column. It is likely that it should have datatype of `category`.\n\nFirst, check the unique values of this column."],"metadata":{}},{"cell_type":"code","source":["mad_men_show_data_pdf.status.value_counts()"],"metadata":{},"outputs":[],"execution_count":37},{"cell_type":"markdown","source":["Now change the datatype for the `status` column with the `dtype` parameter."],"metadata":{}},{"cell_type":"code","source":["%python\nimport numpy as np\nmad_men_show_data_pdf = pd.read_csv('/dbfs/mnt/datalab-datasets/538_data/mad-men/show-data.csv', \n                                    names=['performer', 'show', 'show_start', 'show_end',\n                                           'status', 'char_end', 'years_since', 'lead', \n                                           'support', 'shows', 'score', 'score_y',\n                                           'lead_notes', 'support_notes', 'show_notes'],\n                                    dtype={'show_start': np.float32, \n                                           'show_end': np.float32,\n                                           'score_y': np.float32,\n                                           'status': 'category'},\n                                    header=0, \n                                    na_values=['PRESENT','#DIV/0!'],\n                                    encoding = 'ISO-8859-1')\nmad_men_show_data_pdf \\\n  .info()"],"metadata":{},"outputs":[],"execution_count":39},{"cell_type":"markdown","source":["Great. Now, all columns really are read with acceptable datatypes."],"metadata":{}},{"cell_type":"markdown","source":["## 3. `read_excel`"],"metadata":{}},{"cell_type":"markdown","source":["List the spreadsheets in the `per_diem` directory."],"metadata":{}},{"cell_type":"code","source":["%sh ls -hot /dbfs/mnt/datalab-datasets/per_diem/*.xls"],"metadata":{},"outputs":[],"execution_count":43},{"cell_type":"markdown","source":["The following command uses the pandas function `read_excel` to read in the spreadsheet."],"metadata":{}},{"cell_type":"code","source":["%python\nimport pandas as pd\nmay_2017_pdf = pd.read_excel('/dbfs/mnt/datalab-datasets/per_diem/May2017PD.xls')\nmay_2017_pdf.info()"],"metadata":{},"outputs":[],"execution_count":45},{"cell_type":"markdown","source":["Notice the column names (they are in mixed case and contain spaces) and notice the columns with type `object`. \n\nWe fix the column names with the `names` parameter."],"metadata":{}},{"cell_type":"code","source":["%python\nimport pandas as pd\nmay_2017_pdf = pd.read_excel('/dbfs/mnt/datalab-datasets/per_diem/May2017PD.xls',\n                             names=['country_name', 'location', 'season_code', 'season_start_date', \n                                    'season_end_date', 'lodging', 'meals_incidentals', 'per_diem', \n                                    'effective_date', 'footnote_reference', 'location_code']\n                            )\nmay_2017_pdf.info()"],"metadata":{},"outputs":[],"execution_count":47},{"cell_type":"code","source":["may_2017_pdf \\\n  .select_dtypes(include=['object']) \\\n  .head()"],"metadata":{},"outputs":[],"execution_count":48},{"cell_type":"markdown","source":["The first three columns should have datatype `category`. Make is so number 1."],"metadata":{}},{"cell_type":"code","source":["%python\nimport pandas as pd\nmay_2017_pdf = pd.read_excel('/dbfs/mnt/datalab-datasets/per_diem/May2017PD.xls',\n                             names=['country_name', 'location', 'season_code', 'season_start_date', \n                                    'season_end_date', 'lodging', 'meals_incidentals', 'per_diem', \n                                    'effective_date', 'footnote_reference', 'location_code'],\n                             dtypes={'country_name': 'category',\n                                     'location': 'category',\n                                     'season_code': 'category'}\n                            )\nmay_2017_pdf.info()"],"metadata":{},"outputs":[],"execution_count":50},{"cell_type":"code","source":["%python\nimport pandas as pd\nmay_2017_pdf = pd.read_excel('/dbfs/mnt/datalab-datasets/per_diem/May2017PD.xls',\n                             names=['country_name', 'location', 'season_code', 'season_start_date', \n                                    'season_end_date', 'lodging', 'meals_incidentals', 'per_diem', \n                                    'effective_date', 'footnote_reference', 'location_code'],\n                             dtype={'country_name': 'category',\n                                     'location': 'category',\n                                     'season_code': 'category'}\n                            )\nmay_2017_pdf.info()"],"metadata":{},"outputs":[],"execution_count":51},{"cell_type":"markdown","source":["Checking this message on Google reveals that we should use the `converters` parameter. See [stackoverflow](https://stackoverflow.com/questions/32591466/python-pandas-how-to-specify-data-types-when-reading-an-excel-file).\n\nFurther checking [the documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_excel.html) reveals that the `dtype` parameter was added in version `0.20.0` and we are running `0.19.2`. (See below.)"],"metadata":{}},{"cell_type":"code","source":["pd.__version__"],"metadata":{},"outputs":[],"execution_count":53},{"cell_type":"markdown","source":["There seems to be little documentation on the `converters` parameter.  Because of this I use the `assign` method to change the type of the column after reading the spreadsheet with `read_excel` as follows."],"metadata":{}},{"cell_type":"code","source":["%python\nimport pandas as pd\nmay_2017_pdf = \\\npd.read_excel('/dbfs/mnt/datalab-datasets/per_diem/May2017PD.xls',\n              names=['country_name', 'location', 'season_code', 'season_start_date', \n                     'season_end_date', 'lodging', 'meals_incidentals', 'per_diem', \n                     'effective_date', 'footnote_reference', 'location_code']\n             ) \\\n  .assign(country_name=lambda df: df.country_name.astype('category'),\n          location    =lambda df: df.location.astype('category'),\n          season_code =lambda df: df.season_code.astype('category')\n         )\n        \nmay_2017_pdf.info()"],"metadata":{},"outputs":[],"execution_count":55},{"cell_type":"markdown","source":["See the documentation on the `assign` method:\n- https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.assign.html"],"metadata":{}},{"cell_type":"markdown","source":["Make one last check on the columns of type `object`."],"metadata":{}},{"cell_type":"code","source":["%python\nmay_2017_pdf \\\n  .select_dtypes(include=['object']) \\\n  .head()"],"metadata":{},"outputs":[],"execution_count":58},{"cell_type":"markdown","source":["That looks like a reasonable `object` column. The `may_2017_pdf` dataframe is ready to be analyzed."],"metadata":{}},{"cell_type":"markdown","source":["## 4. `read_json`"],"metadata":{}},{"cell_type":"markdown","source":["List the JSON files in the `/dbfs/mnt/datalab-datasets/JSON` directory."],"metadata":{}},{"cell_type":"code","source":["%sh ls -hot /dbfs/mnt/datalab-datasets/JSON"],"metadata":{},"outputs":[],"execution_count":62},{"cell_type":"markdown","source":["The first example uses the `zips.json` file. It only contains three records, which are displayed below."],"metadata":{}},{"cell_type":"code","source":["%sh head -n 3 /dbfs/mnt/datalab-datasets/JSON/zips.json"],"metadata":{},"outputs":[],"execution_count":64},{"cell_type":"markdown","source":["Read the file using `read_json`. The `lines=True` parameter indicates that there should be one JSON object per line. These JSON objects correspond to rows/records. The key names correspond to column names."],"metadata":{}},{"cell_type":"code","source":["%python\nzips_pdf = pd.read_json('/dbfs/mnt/datalab-datasets/JSON/zips.json',\n                        lines=True)\nzips_pdf.info()"],"metadata":{},"outputs":[],"execution_count":66},{"cell_type":"markdown","source":["List the first five records of the dataframe. Notice the `loc` column. It seems to contain a list, but has the `object` datatype."],"metadata":{}},{"cell_type":"code","source":["zips_pdf.head()"],"metadata":{},"outputs":[],"execution_count":68},{"cell_type":"code","source":["%python \nzips_pdf.loc"],"metadata":{},"outputs":[],"execution_count":69},{"cell_type":"markdown","source":["The `loc` indexer masks the `loc column` so it should be renamed. The `read_json` function has no `names` parameter so we use the `rename` method. See the documentation for details on the `rename` method: \n- https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.rename.html"],"metadata":{}},{"cell_type":"code","source":["%python\nzips_pdf = \\\npd.read_json('/dbfs/mnt/datalab-datasets/JSON/zips.json',\n             lines=True) \\\n  .rename(columns={'loc':'lon_lat'})\nzips_pdf.info()"],"metadata":{},"outputs":[],"execution_count":71},{"cell_type":"markdown","source":["The `city` and `state` columns should have datatype `category`. The `dtype` parameter didn't work so the `assign` method is used."],"metadata":{}},{"cell_type":"code","source":["%python\nzips_pdf = \\\npd.read_json('/dbfs/mnt/datalab-datasets/JSON/zips.json',\n             lines=True) \\\n  .rename(columns={'loc':'lon_lat'}) \\\n  .assign(city =lambda df: df.city .astype('category'),\n          state=lambda df: df.state.astype('category')\n         )\n\nzips_pdf.info()"],"metadata":{},"outputs":[],"execution_count":73},{"cell_type":"markdown","source":["The `lon_lat` column should be split into two columns, but that will have to wait until later."],"metadata":{}},{"cell_type":"markdown","source":["__The End__"],"metadata":{}}],"metadata":{"name":"1.4 Read data files","notebookId":569001},"nbformat":4,"nbformat_minor":0}